### ISSUE TITLE
onSelectedRowsChange called by selectableRowSelected, infinite loop

### ISSUE THREAD

**Issue by yuvachang** [jbetancur/react-data-table-component#930]
## Describe the bug
Coming from v6.11.5 to v7+. 
No issue in v6.

My _onSelectedRowsChange_ callback updates the data--part of what it does adds "isSelected: true" to the selected row. 
I then use _selectableRowSelected = (row) => row.isSelected_ to keep that row selected when component updates with the new data. 

However, right now, the _onSelectedRowsChange_ callback is called on mount with _selectedCount === 0_ if a _selectableRowSelected_ callback is provided to the DataTable component. 

Infinite loop occurs if I select a row and _onSelectedRowsChange_ updates data. The _onSelectedRowsChange_ to get called with _selectedCount === 0_ and _selectedCount === 1_ over and over. 

I hope this is makes sense. v6 the _selectableRowSelected_ callback did not cause any extra _onSelectedRowsChange_ calls, and I'm wondering if this is an intentional change. If it is, is there a 'correct' way of updating Table data while keeping rows selected? 

I am passing the data from parent component as suggested: 
`<Table cart rowData={lines} />`

My DataTable in the Table component looks like this:
```
<DataTable
          columns={this.props.columns}
          dense
          data={this.props.rowData}
          noDataComponent={noDataComponent}
          pagination
          paginationPerPage={50}
          paginationRowsPerPageOptions={[10, 25, 50]}
          persistTableHead
          onSelectedRowsChange={this.handleSelectRow}
          selectableRows={selectableRows}
          selectableRowSelected={(row) => row.isSelected} // This causes the infinite loop, removing it stops the loop but won't render row as selected when the data updates.
          selectableRowDisabled={(row) => row.disableSelect}
          selectableRowsNoSelectAll={false}
          selectableRowsHighlight
          selectableRowsVisibleOnly={false}
          striped
/>
```

## Expected behavior
onSelectedRowsChange callback ONLY called when user selects a row, not onMount of component/Table instance.

## Versions (please complete the following information)
 - React: 16.13.1
 - Styled Components: 5.2.0
 - OS: Ubuntu 20.04
 - Browser: Chrome 94.0.4606.54

I know I'm probably not providing much, but please let me know what else I can share.
Thanks!

**Comment by jbetancur** [comment]
This is definitely due to an internal change in V7 and I was able to duplicate an infinite loop if you try to to update state with the selectedRows provided in `onSelectedRowsChange`.

I'm going to look into why this is the case though but might take me some time to troubleshoot what changed.  

How are you updating `isSelected`? Are you making an API call out to do this? If so the result of that API  updating state in its own effect call would then update `data` instead of via `onSelectedRowsChange`. RDT `selectableRowSelected` would then maintain the selected state without having to change the local `data` state.

What is your `this.handleSelectRow` code doing exactly?

**Comment by yuvachang** [comment]
Hey, thanks for speedy response, I appreciate the help!

My `this.handleSelectRow` is calling a Redux action (albeit no external API) that is also updating other components to respond to selected row(s) data. However, I am simultaneously reusing the Table component for different sets of data, so  my `this.handleSelectRow` is updating the `isSelected` fields to track what's been selected in each set of data (users need to switch between data sets but keep previously selected rows). 
-- On v6, I believe this is behaving as intended because `onSelectedRowsChange` is triggered only once and selectableRowSelected is tracking row selection correctly with only one component update. 
^ Is there a better/intended way of handling this (tracking selected rows on each data set) or am I practicing some anti-pattern?

Thanks again!! This is an amazing library and has been a lifesaving crux of my project.

**Comment by jbetancur** [comment]
Thank you! Hopefully I will have a fix for this soon.

**Comment by jbetancur** [comment]
Fixed in 7.4.2 - should be available shortly

**Comment by jbetancur** [comment]
7.4.2 is now available!

**Comment by smartin36** [comment]
Thank you!

**Comment by mayssalns** [comment]
> 7.4.2 is now available!

The infinite loop still persists in versions v7.4.6, and v7.5, the function runs without the click action, causing an infinite loop.

v7.4.2 It works

**Comment by PedroSass** [comment]
> > 7.4.2 is now available!
> 
> The infinite loop still persists in versions v7.4.6, and v7.5, the function runs without the click action, causing an infinite loop.
> 
> v7.4.2 It works

Thanks for the info, I didn't think downgrading would solve this problem. (Y)

**Comment by CatPerry** [comment]
I know this issue is closed, but for anyone who lands here—and they will if using v7 of React Table...

None of the above (or anywhere else on the interwebs) worked for me but here's what did. And I imagine, with the number of people who've asked this same question about React Table v7, that in version 8 of React Table, the state, especially state having to do with selection, has been ironed out. But anyway this is the ONLY workaround that worked for me, loosely based on this [codesandbox][1]. 

// Parent component
1) I used a `useCallback` hook to prevent my `setSelectedRows` (your `setSelection`) function being passed to the child as a new function and kicking off multiple renders. 

2) And I passed in an initial state that sets React Table's `selectedRowIds` with the parent's `selectedRows`:


```
const ParentDoingParentyThings = () => {
  const [selectedRows, setSelectedRows] = React.useState({});
  const handleSelection = React.useCallback((value) => {
    setSelectedRows(value);   
    }, []); 

return (
  <Table
        data={tableData}
        setSelectedRows={handleSelection}
        initialState={{ sortBy: [{ id: 'totalSeries', desc: true }], selectedRowIds: selectedRows }}
        ...
      />
  )
}
```
// And the Table / child
Then!... 
1) I use this lesser know function from React Table called `useMountedLayoutEffect`. Dont ask me in detail what it does, but it works, that's what it does.... 
2) Then I pull from RT's `selectedRowIds` state
3) And I use RT's `useMountedLayoutEffect` function like I would a `useEffect`

```
import {
  ...
  useMountedLayoutEffect,
} from 'react-table';

export const Table = ({
  data,
  setSelectedRows,
  initialState,
  ...
}: Props) => {
  ...
  const {
    getTableProps,
    getTableBodyProps,
    headerGroups,
    rows,
    prepareRow,
    state: { selectedRowIds },
  } = useTable(
    {
      data: tableData,
      initialState,
      ...
    },
    useSortBy,
    useRowSelect,
  );

  useMountedLayoutEffect(() => {
    setSelectedRows && setSelectedRows(selectedRowIds);
  }, [setSelectedRows, selectedRowIds]);

  return (
    <table {...getTableProps()} className={className}>
      ...
    </table>
```


  [1]: https://codesandbox.io/embed/tannerlinsleyreact-table-row-selection-controlled-euyku?fontsize=14&hidenavigation=1&theme=dark

**Comment by rajat-est** [comment]
@jbetancur This issue is still present. Could you please review and update ?

**Comment by rahulsrinivasan1** [comment]
@jbetancur Any update on the bug as even i am facing the same issue

**Comment by gioserbug** [comment]
@jbetancur any update about this bug? thanks!

**Comment by rajat22999** [comment]
@jbetancur Yes, Please look into this. I'm still facing this issue. Hope you fix it soon.

**Comment by musharafshaikh16** [comment]
Same here issue still exist!

**Comment by M-Umais-Hassan** [comment]
I think this is a kind of limitation or may be react-data-table-component team want us not to bind the table 2 way for selection purpose and the table will handle this on its own. 
Two way binding happens when we have a value prop e.g. selectableRowSelected and other thing we have is value change handler in our case it is "onSelectedRowsChange".

But the problem is, both of them are functions while ideally the value has not to be a function. I think this is the reason why infinite loop is causing.

So, I have solved it something like this, although it is not a good way but it will work,

```
const firstRender = useRef(true);

const selectProps = {
    ...(firstRender.current && {
      selectableRowSelected: () => preSelectRows(selected),
    }),
    ...(!firstRender.current && {
      onSelectedRowsChange: handleSelection,
    }),
  };

useEffect(() => firstRender.current = false, []);

<GenericDataTable
          {...selectProps}
          noDataComponent={<NoDataComponent />}
          persistTableHead={true}
          columns={columns}
          data={outletsListingData?.outlets}
          loading={loading}
          selectableRows={true}
          keyField="outlet_name_en"
/>
```

Any of your remarks are highly appreciated. Again that is not a valid solution, but it works for now.

**Comment by BagusHartiansyah-2G18** [comment]
I saw the enthusiasm of my friends in finding a solution, finally I wanted to take part.

I have a method that starts from:
1. create a variable count (state).
2. method rowSelectCritera with action :
_count((count+1)); 
return row.selected;
3. Pay attention to the rendering update,  I gave the value in the table 
rowSelectCritera={(count<dmember ? rowSelectCritera: false)} 

hopefully can help


saya melihat semangat dari teman - teman dalam mencari solusinya, akhirnya saya ingin ikut andil.
saya mempunyai cara yakni dimulai dari :
1. membuat variable count (state).
2. method  rowSelectCritera dengan  action :
_count((count+1)); 
return row.selected;
3.  perhatikan update rendernya, saya memberi nilai pada tabel nya 
rowSelectCritera={(count<dmember ? rowSelectCritera: false)} 

semoga dapat membantu
